File: src/de/stephanlindauer/criticalmaps/model/ChatModel.java
package de.stephanlindauer.criticalmaps.model;

import com.crashlytics.android.Crashlytics;
import de.stephanlindauer.criticalmaps.vo.chat.IChatMessage;
import de.stephanlindauer.criticalmaps.vo.chat.OutgoingChatMessage;
import de.stephanlindauer.criticalmaps.vo.chat.ReceivedChatMessage;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.util.*;

public class ChatModel {

    private ArrayList<ReceivedChatMessage> chatMessages = new ArrayList<ReceivedChatMessage>();
    private ArrayList<OutgoingChatMessage> outgoingMassages = new ArrayList<>();

    //singleton
    private static ChatModel instance;

    public static ChatModel getInstance() {
        if (ChatModel.instance == null) {
            ChatModel.instance = new ChatModel();
        }
        return ChatModel.instance;
    }

    public void setNewJson(JSONObject jsonObject) throws JSONException, UnsupportedEncodingException {
        if (chatMessages == null) {
            chatMessages = new ArrayList<ReceivedChatMessage>();
        } else {
            chatMessages.clear();
        }

        Iterator<String> keys = jsonObject.keys();
        while (keys.hasNext()) {
            String key = keys.next();
            JSONObject value = jsonObject.getJSONObject(key);
            String message = URLDecoder.decode(value.getString("message"), "UTF-8");
            Date timestamp = new Date(Long.parseLong(value.getString("timestamp")) * 1000);
            String identifier = key;

            for (int i = outgoingMassages.size() - 1; i > -1; i--) {
                OutgoingChatMessage outgoingMessageToMaybeDelete = outgoingMassages.get(i);
                if (outgoingMessageToMaybeDelete.getIdentifier().equals(identifier)) {
                    outgoingMassages.remove(i);
                }
            }

            chatMessages.add(new ReceivedChatMessage(message, timestamp));
        }

        Collections.sort(chatMessages, new Comparator<ReceivedChatMessage>() {
            @Override
            public int compare(ReceivedChatMessage oneChatMessages, ReceivedChatMessage otherChatMessage) {
                return oneChatMessages.getTimestamp().compareTo(otherChatMessage.getTimestamp());
            }
        });
    }

    public void setNewOutgoingMessage(OutgoingChatMessage newOutgoingMessage) {
        outgoingMassages.add(newOutgoingMessage);
    }

    public JSONObject getOutgoingMessagesAsJson(String uniqueDeviceIdHashed) {
        JSONObject jsonObject = new JSONObject();
        Integer counter = 0;

        for (int i = 0; i < outgoingMassages.size(); i++) {
            OutgoingChatMessage outgoingChatMessage = outgoingMassages.get(i);
            try {
                JSONObject messageObject = new JSONObject();
                messageObject.put("text", outgoingChatMessage.getUrlEncodedMessage());
                messageObject.put("timestamp", outgoingChatMessage.getTimestamp().getTime());
                messageObject.put("uniqueDeviceIdHashed", uniqueDeviceIdHashed);
                messageObject.put("identifier", outgoingChatMessage.getIdentifier());
                jsonObject.put(counter.toString(), messageObject);
                counter++;
            } catch (JSONException e) {
                Crashlytics.logException(e);
            }
        }
        return jsonObject;
    }

    public ArrayList<IChatMessage> getSavedAndOutgoingMessages() {
        ArrayList<IChatMessage> mergeArrayList = new ArrayList<IChatMessage>();
        mergeArrayList.addAll(chatMessages);
        mergeArrayList.addAll(outgoingMassages);
        return mergeArrayList;
    }

    public boolean hasOutgoingMessages() {
        return outgoingMassages.size() > 0;
    }
}

File: app/src/main/java/de/stephanlindauer/criticalmaps/model/ChatModel.kt
package de.stephanlindauer.criticalmaps.model

import de.stephanlindauer.criticalmaps.model.chat.ReceivedChatMessage
import de.stephanlindauer.criticalmaps.utils.AeSimpleSHA1
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import timber.log.Timber
import java.io.UnsupportedEncodingException
import java.net.URLDecoder
import java.net.URLEncoder
import java.util.Date
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ChatModel @Inject constructor(
    private val userModel: UserModel
) {
    private var receivedChatMessages = mutableListOf<ReceivedChatMessage>()

    fun getReceivedChatMessages(): List<ReceivedChatMessage> = receivedChatMessages

    @Throws(JSONException::class, UnsupportedEncodingException::class)
    fun setFromJson(jsonArray: JSONArray) {
        receivedChatMessages = ArrayList(jsonArray.length())

        for (i in 0 until jsonArray.length()) {
            val jsonObject = jsonArray.getJSONObject(i)

            val device = URLDecoder.decode(jsonObject.getString("device"), "UTF-8")
            val identifier = URLDecoder.decode(jsonObject.getString("identifier"), "UTF-8")
            val message = URLDecoder.decode(jsonObject.getString("message"), "UTF-8")
            val timestamp = Date(jsonObject.getString("timestamp").toLong() * 1000)

            receivedChatMessages.add(ReceivedChatMessage(message, timestamp))
        }

        receivedChatMessages.sortBy { it.timestamp }
    }

    fun createNewOutgoingMessage(message: String): JSONObject {
        val messageObject = JSONObject()
        try {
            messageObject.put("text", urlEncodeMessage(message))
            messageObject.put("identifier", AeSimpleSHA1.SHA1(message + Math.random()) ?: "")
            messageObject.put("device", userModel.changingDeviceToken)
        } catch (e: JSONException) {
            Timber.d(e)
        }
        return messageObject
    }

    private fun urlEncodeMessage(messageToEncode: String): String {
        return try {
            URLEncoder.encode(messageToEncode, "UTF-8")
        } catch (e: UnsupportedEncodingException) {
            ""
        }
    }

    companion object {
        const val MESSAGE_MAX_LENGTH = 255
    }
}
